src/neo4j_graphrag/tool.py

```diff
@@ -0,0 +1,261 @@
+from abc import ABC
+from enum import Enum
+from typing import Any, Dict, List, Callable, Optional, Union, ClassVar
+from pydantic import BaseModel, Field, model_validator
+
+
+class ParameterType(str, Enum):
+    """Enum for parameter types supported in tool parameters."""
+
+    STRING = "string"
+    INTEGER = "integer"
+    NUMBER = "number"
+    BOOLEAN = "boolean"
+    OBJECT = "object"
+    ARRAY = "array"
+
+
+class ToolParameter(BaseModel):
+    """Base class for all tool parameters using Pydantic."""
+
+    description: str
+    required: bool = False
+    type: ClassVar[ParameterType]
+
+    def model_dump_tool(self) -> Dict[str, Any]:
+        """Convert the parameter to a dictionary format for tool usage."""
+        result = {"type": self.type, "description": self.description}
+        return result
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "ToolParameter":
+        """Create a parameter from a dictionary."""
+        param_type = data.get("type")
+        if not param_type:
+            raise ValueError("Parameter type is required")
+
+        # Find the appropriate class based on the type
+        param_classes = {
+            ParameterType.STRING: StringParameter,
+            ParameterType.INTEGER: IntegerParameter,
+            ParameterType.NUMBER: NumberParameter,
+            ParameterType.BOOLEAN: BooleanParameter,
+            ParameterType.OBJECT: ObjectParameter,
+            ParameterType.ARRAY: ArrayParameter,
+        }
+
+        param_class = param_classes.get(param_type)
+        if not param_class:
+            raise ValueError(f"Unknown parameter type: {param_type}")
+
+        # Use type ignore since mypy doesn't understand dynamic class instantiation
+        return param_class.model_validate(data)  # type: ignore
+
+
+class StringParameter(ToolParameter):
+    """String parameter for tools."""
+
+    type: ClassVar[ParameterType] = ParameterType.STRING
+    enum: Optional[List[str]] = None
+
+    def model_dump_tool(self) -> Dict[str, Any]:
+        result = super().model_dump_tool()
+        if self.enum:
+            result["enum"] = self.enum
+        return result
+
+
+class IntegerParameter(ToolParameter):
+    """Integer parameter for tools."""
+
+    type: ClassVar[ParameterType] = ParameterType.INTEGER
+    minimum: Optional[int] = None
+    maximum: Optional[int] = None
+
+    def model_dump_tool(self) -> Dict[str, Any]:
```

Comment:
Is `model_dump_tool` only used to discard null values? If so, there is an option in `model_dump` for this (`exclude_none=True` by memory)

Reply:
I know I'm a bit nitpicking here, but I've started to work on the implementation for Vertex AI, and some parameters need to be excluded, so if we can rely on Pydantic for this instead of manually re-implementing it, that would be great :) but if it's not possible, it's not possible.

Reply:
It does a bit more than that mapping some names "min_items" -> "minItems", but that could be removed by naming the parameters "minItems" directly. 
Your call, do you want me to change it?

Reply:
Ok I see, no do not change anything now, we'll add improvements as we need them. They are unit tested so that's great :)